File: 프로그래머스/2/42584. 주식가격/주식가격.py
# 이 문제도 인덱스를 다루는 것이 핵심

def solution(prices):
    # 결과를 저장할 리스트를 0으로 초기화
    answer = [0] * len(prices)
    # 가격의 인덱스를 저장할 스택
    stack = []

    # 모든 가격을 순회
    for i in range(len(prices)):
        # 현재 가격이 스택의 마지막 가격보다 작다면, 가격이 떨어진 시점
        while stack and prices[stack[-1]] > prices[i]:
            top = stack.pop()  # 스택에서 인덱스를 꺼내고
            answer[top] = i - top  # 현재 인덱스에서 top의 인덱스를 빼기
        
        # stack이 비어 있으면 추가
        stack.append(i)
    
    # 루프가 끝난 이후 남아 있는 스택 처리
    for top in stack:
         # 끝까지 유지된 기간 계산 (len(prices)-1 은 한 몸이다.)
        answer[top] = len(prices) - 1 - top 

    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.ArrayDeque;

class Solution {
    public static int[] solution(int[] prices) {
        int n = prices.length;
        int[] answer = new int[n];
        ArrayDeque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            // 스택이 비어있지 않고, 현재 가격이 스택의 맨 위 가격보다 작을 때
            while (!stack.isEmpty() && prices[stack.peekLast()] > prices[i]) {
                // 스택의 맨 위 가격 인덱스를 꺼내서
                int top = stack.pollLast();
                // 현재 인덱스 i에서 스택의 맨 위 가격 인덱스 top을 빼면 가격이 떨어지지 않은 기간이 됨
                answer[top] = i - top;
            }
            // 현재 가격의 인덱스를 스택에 추가
            stack.offerLast(i);
        }

        // 마지막 가격은 가격이 떨어지지 않은 기간이 없으므로 0으로 설정
        while (!stack.isEmpty()) {
            int top = stack.pollLast();
            answer[top] = n - 1 - top;
        }

        return answer;
    }
}